packages/scm-extra/src/browser/history/scm-history-widget.tsx

```diff
@@ -15,92 +15,158 @@
+    protected listView: ScmHistoryList | undefined;
     protected hasMoreCommits: boolean;
     protected allowScrollToSelected: boolean;
 
     protected status: {
         state: 'loading',
     } | {
         state: 'ready',
-        commits: GitCommitNode[];
+        commits: ScmCommitNode[];
     } | {
         state: 'error',
         errorMessage: React.ReactNode
     };
 
+    protected readonly toDisposeOnRepositoryChange = new DisposableCollection();
+
+    protected historySupport: ScmHistorySupport | undefined;
+
     constructor(
+        @inject(ScmService) protected readonly scmService: ScmService,
         @inject(OpenerService) protected readonly openerService: OpenerService,
-        @inject(GitCommitDetailOpenHandler) protected readonly detailOpenHandler: GitCommitDetailOpenHandler,
         @inject(ApplicationShell) protected readonly shell: ApplicationShell,
         @inject(FileSystem) protected readonly fileSystem: FileSystem,
-        @inject(Git) protected readonly git: Git,
         @inject(ScmAvatarService) protected readonly avatarService: ScmAvatarService,
         @inject(WidgetManager) protected readonly widgetManager: WidgetManager,
-        @inject(GitDiffContribution) protected readonly diffContribution: GitDiffContribution) {
+    ) {
         super();
-        this.id = GIT_HISTORY_ID;
-        this.scrollContainer = 'git-history-list-container';
-        this.title.label = GIT_HISTORY_LABEL;
-        this.title.caption = GIT_HISTORY_LABEL;
-        this.title.iconClass = 'fa git-history-tab-icon';
+        this.id = SCM_HISTORY_ID;
+        this.scrollContainer = 'scm-history-list-container';
+        this.title.label = SCM_HISTORY_LABEL;
+        this.title.caption = SCM_HISTORY_LABEL;
+        this.title.iconClass = 'fa scm-history-tab-icon';
         this.title.closable = true;
-        this.addClass('theia-git');
+        this.addClass('theia-scm');
+        this.addClass('theia-scm-history');
         this.resetState();
         this.cancelIndicator = new CancellationTokenSource();
     }
 
     @postConstruct()
     protected init(): void {
+        this.refreshOnRepositoryChange();
+        this.toDispose.push(this.scmService.onDidChangeSelectedRepository(() => this.refreshOnRepositoryChange()));
         this.toDispose.push(this.labelProvider.onDidChange(event => {
-            if (this.gitNodes.some(node => GitFileChangeNode.is(node) && event.affects(new URI(node.uri)))) {
+            if (this.scmNodes.some(node => ScmFileChangeNode.is(node) && event.affects(new URI(node.fileChange.uri)))) {
                 this.update();
             }
         }));
     }
 
+    protected refreshOnRepositoryChange(): void {
+        this.toDisposeOnRepositoryChange.dispose();
+
+        const repository = this.scmService.selectedRepository;
+        if (repository) {
+            this.historySupport = repository.provider.get && repository.provider.get(ScmHistorySupport);
```

Comment:
Why cannot we introduce a new interface like:
```ts
interface ScmHistoryProvider extends ScmProvider {
    historySupport?: ScmHistorySupport
}
namespace ScmHistoryProvider {
    export function is(scmProvider: ScmProvider | undefined): provider is ScmHistoryProvider {
        retunrn scmProvider && 'historySupport' in scmProvider;
    }
}
```

so here we just do down cast check without passing containers around and providing unused by scm get method to lookup random service?

Reply:
That won't work because two different packages might both be providing a custom API implementation to an SCM.  Currently ScmHistorySupport is the only extension API.  But suppose there was another extension API.  For example, suppose we decide to move the Amend support out of the scm package into yet another package.  Or suppose an extender consumes another custom API.  The provider can't implement all those interfaces (ScmHistoryProvider, ScmAmendProvider, etc) when the definitions of those interfaces come from different packages that may or may not be installed.

This is support for binding with a scope of SCM type.  It allows Alice to define an interface with the functions she needs to implement her SCM view.  It allows Bob to bind an implementation of that interface to the Git container.  It allows Chris to bind an implementation of that interface to the SVN container.  All are doing so in their own packages (not changing scm or scm-extra packages).

Reply:
On thinking more about this, I may be able to come up with something better.  It will be tricky to inject implementations into the SCM type child container from a different package than the one that added the SCM type.  I may push some changes tomorrow.

Reply:
> The provider can't implement all those interfaces (ScmHistoryProvider, ScmAmendProvider, etc) when the definitions of those interfaces come from different packages that may or may not be installed.

Should not provider depends on these packages if he is going to implement them? I think it is what we have now, git extension depends on scm and scm-extra and can implement both interfaces, it will depends on scm-amend if it is required too.

if a provider only interested in type he has to add these packages as dev dependencies and import only types. In this case scm-extra and scm-amend don't come with git extension by default.

Reply:
> Should not provider depends on these packages if he is going to implement them? I think it is what we have now, git extension depends on scm and scm-extra and can implement both interfaces, it will depends on scm-amend if it is required too.

I anticipated that it will be useful to add a new API and add a Git implementation without needing to modify the git package.  However that is not essential for us so I made the changes to simply get the history support directly from the provider.

The problem is, however, that the provider cannot depend on the history support.  It cannot have history support injected into it because the provider is injected into the history support.  Furthermore git-scm-provider.ts cannot have anything imported from git-history-support because that would also be a circular dependency.  Removing the provider as a dependency to git-history-support would not be good enough because git-history-support depends on GitRepositoryTracker which depends on GitRepositoryProvider.  So I was right back to removing all the history imports from git-scm-provider.

ScmFactory has been removed.  The scm-type containers have been removed and scm-input has been reverted.  The child container bindings have been moved out of git-scm-provider so it can bind history stuff.  You will see in that binding code a small hack to set the history support into the provider.

Reply:
ok, thanks, could we remove `ScmProvider.get` please?